R/dm_wrap.R

```diff
@@ -119,37 +119,50 @@ dm_wrap_tbl_plan <- function(dm, root) {
 #' dm_nrow(roundtrip)
 dm_unwrap_tbl <- function(dm, ptype) {
   check_dm(ptype)
-  # unwrap all tables and their unwrapped children/parents
-  unwrapped_table_names <- character(0)
-  repeat {
-    to_unwrap <- setdiff(names(dm), unwrapped_table_names)[1]
-    done_unwrapping <- is.na(to_unwrap)
-    if (done_unwrapping) break
-    dm <- dm_unwrap_tbl1(dm, !!to_unwrap, ptype)
-    unwrapped_table_names <- c(unwrapped_table_names, to_unwrap)
-  }
-  dm
+
+  unwrap_plan <- imap_dfr(dm_get_tables(dm), dm_unwrap_tbl_plan)
+
+  unwrapped_dm <- reduce(
+    transpose(unwrap_plan),
+    function(dm, row) {
+      exec(row$action, dm, row$table, row$col, ptype)
+    },
+    .init = dm
+  )
+  unwrapped_dm
 }
 
-dm_unwrap_tbl1 <- function(dm, table, ptype) {
-  # process args and build names
-  table_name <- dm_tbl_name(dm, {{ table }})
-  table <- dm_get_tables_impl(dm)[[table_name]]
+dm_unwrap_tbl_plan <- function(table, table_name) {
   nms <- names(table)
 
-  # detect parent and children tables
   children <- nms[map_lgl(table, inherits, "nested")]
   parents <- nms[map_lgl(table, inherits, "packed")]
 
-  # unnest children tables
-  for (child_name in children) {
-    dm <- dm_unnest_tbl(dm, !!table_name, col = !!child_name, ptype = ptype)
-  }
+  unnest_plan <-
+    tibble(
+      action = "dm_unnest_tbl",
+      table = table_name,
+      col = children
+    )
 
-  # unpack parent tables
-  for (parent_name in parents) {
-    dm <- dm_unpack_tbl(dm, !!table_name, col = !!parent_name, ptype = ptype)
-  }
+  unpack_plan <-
+    tibble(
+      action = "dm_unpack_tbl",
+      table = table_name,
+      col = parents
+    )
+
+  unwrap_plan_from_children <-
+    # note: we cannot use bind_rows() because of https://github.com/tidyverse/dplyr/issues/6447
```

Comment:
```suggestion
    # note: we cannot use bind_rows() because of https://github.com/tidyverse/dplyr/issues/6447,
    # or even vec_rbind() because of https://github.com/r-lib/vctrs/issues/1640
```

Reply:
> `map_dfr()` and `transpose()` are doing a ton of unnecessary work, should we use simple `map()` instead and omit `transpose()` ?

In `imap_fr(dm_get_tables(dm), dm_unwrap_tbl_plan)` `dm_unwrap_tbl_plan()` returns data frames, not just rows. So just using `imap()`and `transpose()` is not enough, I'd need to `split.data.frame()` every input from `imap()` too. Would it be better ?

> I also wonder how this will work for JSON data where we don't have access to the structure of the children, especially when this is a lazy source.

Yes I figured we'd tackle this as a next step. A problem with json is that I believe there are no zero length elements, so no structure to be found in nested elements where we'd have a zero row data frame in the standard version. To avoid collecting everything we might filter on non empty elements and fetch the first.
We can use a generic at this place in the code.